src/engine/SpatialJoin.cpp

```diff
@@ -0,0 +1,295 @@
+#include "SpatialJoin.h"
+#include "util/AllocatorWithLimit.h"
+#include "util/MemorySize/MemorySize.h"
+#include "global/ValueId.h"
+#include "ValuesForTesting.h"
+#include "parser/ParsedQuery.h"
+#include "VariableToColumnMap.h"
+#include "engine/ExportQueryExecutionTrees.h"
+#include "util/GeoSparqlHelpers.h"
+
+SpatialJoin::SpatialJoin(QueryExecutionContext* qec, SparqlTriple triple,
+  std::optional<std::shared_ptr<QueryExecutionTree>> childLeft,
+  std::optional<std::shared_ptr<QueryExecutionTree>> childRight,
+  int maxDist)
+            : Operation(qec) {
+  this->triple = triple;
+  this->maxDist = maxDist;
+  if (childLeft) {
+    this->childLeft = childLeft.value();
+  }
+  if (childRight) {
+    this->childRight = childRight.value();
+  }
+  
+  if (triple._s.isVariable() && triple._o.isVariable()) {
+    leftChildVariable = triple._s.getVariable();
+    rightChildVariable = triple._o.getVariable();
+  } else {
+    AD_THROW("SpatialJoin needs two variables");
+  }
+}
+
+std::shared_ptr<SpatialJoin> SpatialJoin::addChild(
+      std::shared_ptr<QueryExecutionTree> child, Variable varOfChild){
+  if (varOfChild == leftChildVariable) {
+    childLeft = child;
+    return std::make_shared<SpatialJoin>(getExecutionContext(), triple.value(),
+          childLeft, childRight, maxDist);
+  } else if (varOfChild == rightChildVariable) {
+    childRight = child;
+    return std::make_shared<SpatialJoin>(getExecutionContext(), triple.value(),
+          childLeft, childRight, maxDist);
+  } else {
+    AD_THROW("variable does not match");
+  }
+}
+
+bool SpatialJoin::isConstructed() {
+  if (childLeft && childRight) {
+    return true;
+  }
+  return false;
+}
+
+std::vector<QueryExecutionTree*> SpatialJoin::getChildren() {
+  if (!(childLeft || childRight)) {
+    AD_THROW("SpatialJoin needs two variables");
+  }
+
+  return {childLeft.get(), childRight.get()};
+}
+
+
+string SpatialJoin::getCacheKeyImpl() const {
+  if (childLeft && childRight) {
+    std::ostringstream os;
+    os << "SpatialJoin\nChild1:\n" << childLeft->getCacheKey() << "\n";
+    os << "Child2:\n" << childRight->getCacheKey() << "\n";
+    return std::move(os).str();
```

Comment:
The distance must also be part of the cache key, otherwise it won't work correctly.